architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+Review Kubernetes events and logs for more details on the failure.
+```
+
+**Result:**
+
+1. System detects failure during the upgrade process
+1. Automatic restore is initiated using the pre-upgrade backup
+1. All components are restored to their previous state
+1. User is informed of the failure and suggested next steps
+
+**Exceptions:**
+
+1. If the user data backup restoration fails (rare but possible)
+
+#### Scenario 4: Upgrading across multiple versions
+
+**User Story:** As a DevOps engineer, I need to upgrade Radius from an older version to the latest version in a single operation.
+
+**User Experience:**
+
+```bash
+# Check current version (significantly behind latest)
+> rad version
+RELEASE   VERSION   BICEP     COMMIT
+0.40.0    v0.40.0   0.31.93   1dd17270ec9bc9e725f314fa62c249406034edda
+
+# Upgrade directly to latest version
+> rad upgrade kubernetes --version latest
+
+Initiating Radius upgrade from v0.40.0 to v0.44.0 (latest stable)...
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Multiple version jump detected (v0.40.0 → v0.44.0)
+  ✓ Compatible upgrade path confirmed
+  ✓ Database schema changes detected
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✓ Applications Resource Provider
+  ✓ Controller
+Performing post-upgrade verification...
+  ✓ All components healthy
+  ✓ Database schema updated successfully
+
+Upgrade complete! Radius has been successfully upgraded to v0.44.0.
+Note: Your local Radius CLI is still v0.40.0. To upgrade your CLI, download the latest version.
+```
+
+**Result:**
+
+1. System detects a multi-version upgrade scenario
+1. Pre-flight checks validate that direct upgrade is supported
+1. All components are upgraded to the target version
+1. Verification confirms system health with the new version
+
+**Exceptions:**
+
+1. If direct upgrade path isn't supported between versions
+1. If database migrations encounter issues (this may be the case when we introduce Postgres as the data store)
+1. If intermediate upgrades are required first
+
+## Design
+
+### High-Level Design Diagram
+
+```mermaid
+graph TD
+  CLI["Radius CLI (rad upgrade kubernetes)"] -->|Initiates Upgrade| KubernetesAPI["Kubernetes API"]
+  KubernetesAPI -->|Manages Helm Charts| Helm["Helm"]
+  Helm -->|Upgrades Components| ControlPlane["Radius Control Plane"]
+  subgraph ControlPlane
+      UniversalControlPlane["UCP"]
+      DeploymentEngine["Deployment Engine"]
+      ARP["Applications RP"]
+      Controller["Controller"]
+      DynamicRP["Dynamic RP"]
+  end
+  ControlPlane -->|Stores Metadata| Database["etcd"]
+  ControlPlane -->|Interacts with| Dependencies["Dependencies (Dapr, Contour)"]
+  CLI -->|Logs Progress| User["User"]
+  CLI -->|Performs Pre-flight Checks| PreFlight["Pre-flight Checks"]
+  PreFlight -->|Validates| KubernetesAPI
+  CLI -->|Creates User Data Backup| Backup["User Data Backup"]
```

Comment:
Where is the backup stored?

Reply:
Kubernetes objects may be too platform specific. But for now, I believe we are going to keep them in Kubernetes object like a PVC (https://kubernetes.io/docs/concepts/storage/persistent-volumes/)